pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -117,6 +149,24 @@ func initDrainer(r *ReconcileNodeMaintenance, config *rest.Config) error {
 	return nil
 }
 
+func checkLeaseSupported(r *ReconcileNodeMaintenance) error {
```

Comment:
I think the underlying reason you're having to get "creative" with modified fake clients, is that this signature for this method is wrong.

Compare the example I googled for you.... https://sourcegraph.com/github.com/kubernetes/dashboard@f47dc5e48c2cb1d10bc371f11f2efb8fe1c54260/-/blob/src/app/backend/validation/validaterbacstatus.go#L34

Start with:

```suggestion
func checkLeaseSupported(client kubernetes.Interface) (*LeaseStatus, error) {
```

and the rest should follow more cleanly.

Reply:
my problem with the kubernetes fake client is as that the discovery client field (one of the client sets) set was not exported, and that there was no way to set it through the create API. 

checkLeaseSupport needs to establish a read only member of ReconcileNodeMaintenance - isLeaseSupported bool , so it needs ReconcileNodeMaintenance as argument (otherwise I would have to make up a global isLeaseSupported, and that is probably worse). The reason why it is a seperate function and not called from initDrainer:

Reply:
> my problem with the kubernetes fake client is as that the discovery client field (one of the client sets) set was not exported, and that there was no way to set it through the create API

And yet `ValidateRbacStatus` and `TestValidateRbacStatus` manage to do so.  
Take the time to carefully read and understand their code.

> checkLeaseSupport needs to establish a read only member of ReconcileNodeMaintenance"

no, it doesn't. the caller can and should do this which allows `checkLeaseSupported` to be cleaner and operate directly on a client interface (which is easier to fake for testing).

Reply:
I checked your example: ValidateRbacStatus only accesses the Discovery() method of the clientset object, it is therefore sufficient to pass it an interface that implements just that method. However in our tests we need a full fake client that also returns preregistered responses that are set up in the setFakeClient closure, and in addition to that returns the discovery interface - and that is exactly missing from the regular fake clientset in package k8s.io/client-go/kubernetes/fake

Reply:
still i could use that trick you mentioned just for the call of initDrainer and after that switch the client to another fake client for the rest of the test.

Reply:
changed the checkLeaseSupported interface in the latest commit.